pkg/cvo/status_history.go

```diff
@@ -0,0 +1,235 @@
+package cvo
+
+import (
+	"math"
+	"strings"
+
+	"k8s.io/klog/v2"
+
+	configv1 "github.com/openshift/api/config/v1"
+)
+
+const (
+	// maxFinalEntryIndex is an upper bound index. History entries from 0 to maxFinalEntryIndex are given mostImportantWeight.
+	maxFinalEntryIndex = 4
+
+	// mostImportantWeight is given to history entries that should never be removed including the initial entry,
+	// the most recent entries (indices 0..maxFinalEntryIndex), and the most recently completed update.
+	mostImportantWeight = 1000.0
+
+	// interestingWeight is given to the first or last completed update in a given minor. These are interesting
+	// but not critical.
+	interestingWeight = 30.0
+
+	// partialMinorWeight is given to any partial minor updates between a completed minor version transition.
+	partialMinorWeight = 20.0
+
+	// partialZStreamWeight is given to any partial z-stream updates between a completed z-stream version transition.
+	partialZStreamWeight = -20.0
+
+	// sliceIndexWeight when applied will favor more recent updates and avoid ties.
+	sliceIndexWeight = -1.01
+)
+
+// maxHistorySize is set by the function prune and used by all the history pruning functions. Setting and using
+// this local package variable eases unit testing by allowing testing against a smaller history slice.
+var maxHistorySize int
+
+// prune prunes history when at maxSize by ranking each entry using the above defined weights and removing the entry with
+// the lowest rank. maxSize is passed in to allow the use of smaller values which eases unit test.
+func prune(history []configv1.UpdateHistory, maxSize int) []configv1.UpdateHistory {
+	if len(history) <= maxSize {
+		return history
+	}
+	maxHistorySize = maxSize
+
+	idx := getTheMostRecentCompletedEntryIndex(history)
+
+	var reason string
+	lowestRank := math.MaxFloat64
+	var lowestRankIdx int
+
+	for i := range history {
+		rank := 0.0
+		thisReason := ""
+		if isTheInitialEntry(i) {
+			rank = mostImportantWeight
+		}
+		if rank == 0.0 && isAFinalEntry(i) {
```

Comment:
nit: these `if rank == 0.0 &&` could be collapsed into `else if`.  Or instead of a big `else if` chain, you could use [a `switch`][1]:

```go
switch {
case i <= maxFinalEntryIndex || i == mostRecentCompletedEntryIndex || i == maxSize:
  rank = mostImportantWeight
case ...
```

[1]: https://go.dev/ref/spec#Switch_statements

Reply:
I had originally had different reasons for the first three but then realized they should never be pruned so yeah no reason not to collapse them. I avoided `else if`s because in pass reviews I sensed an aversion to them, not necessarily by you,  but I prefer them here too. Regarding the switch option, I believe keeping the functions makes the code both more readable and more maintainable. Won't be surprised if we go back in here and tweak some things.